### PR TITLE
[WIP] Add reference documentation for built-in controllers

### PR DESCRIPTION
**Pull request by sftim**
Relevant to issue #4135.

Add a reference section that lists the built-in controllers and explains what each of them does.

This change fits in with an idea I have to restructure https://kubernetes.io/docs/concepts/workloads/

People developing things to deploy on Kubernetes need to understand that there is (eg) a Deployment resource; it doesn't matter as much _how_ the Deployment resource gets translated into API actions like creating or removing ReplicaSet objects. Some resources happen to involve more than one controller (eg Job with the Job controller and the TTL-after-finished controller).

If you are interested in looking behind the scenes, I think it helps to have all kube-controller-manager controllers listed in one place, no matter what kind of resource they interact with (Endpoint? ReplicaSet? Namespace?)

/sig apps

You can see this change in broader context by viewing the deployment preview for PR #15373 (or looking at the list of files changed there, etc).

Credit to @lavalamp for the excellent [Kubecon talk](https://youtu.be/zCXiXKMqnuE) that helped me a lot with drafting these documents.

### CODE REVIEW

**Comment by sftim** [comment]
@bradtopol I'd welcome your feedback on this.

File: content/en/docs/reference/controllers/horizontal-pod-autoscaler.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  of target specified) across all targeted pods, and produces a ratio used to scale
+  the number of desired replicas.
+
+  If some of the Pod's containers do not have the relevant resource request set,
+  CPU utilization for the pod will not be defined and the autoscaler will
+  not take any action for that metric. See [algorithm
+  details](#algorithm-details) for more information about
+  how the autoscaling algorithm works.
+
+* For per-pod custom metrics, the controller functions similarly to per-pod resource metrics,
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/horizontal-pod-autoscaler.md]: I've seen a number of questions about what "used" means with regards to resource metrics on Slack and the mailing lists. It's probably a good idea to put a note in here explaining what "raw values, not utilization values" means. Perhaps something like "Resource utilization values are calculated from the Pod and Container resource request and limit amounts. Raw utilization values, on the other hand, are queried from the Node's underlying operating system."

File: content/en/docs/reference/controllers/job.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+controller updates the Job object to mark it complete.
+
+{{% /capture %}}
+
+
+{{% capture body %}}
+
+The job controller is built in to kube-controller-manager.
+
+## Pod creation
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/job.md]: I feel the remainder of this information belongs in a new concepts/workloads/job.md document, actually. The vast majority of this information is about the Job object/concept and not specifically about the controller mechanism under the hood, which is what I think a reference/controllers/job.md document should focus on.

File: content/en/docs/reference/controllers/job.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+created the Job yourself, it is up to you to delete it (eg with
+{{< glossary_tooltip term_id="kubectl" >}}).
+
+The activeDeadlineSeconds applies to the duration of the Job, no matter
+how many Pods are created.
+Once a Job reaches `activeDeadlineSeconds`, the controller terminates
+all its Pods and sets the Job's status to `type: Failed` with
+`reason: DeadlineExceeded`.
+
+## Cleanup for finished Jobs
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/job.md]: This section, however, I think belongs in the controller reference doc because it touches on more low-level/implementation information about how the controller cleans up after itself.

File: content/en/docs/reference/controllers/node-lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+## Controller behaviour
+
+This controller observes the behavior of kubelet on a node, and sets (potentially
+also removes) {{< glossary_tooltip text="taints" term_id="taint" >}} on Nodes
+that reflect its findings.
+
+For example: if kubelet stops reporting that a worker node is healthy, the
+controller can apply a taint to prevent scheduling new Pods there.
+
+If there is a significant problem &ndash; maybe the entire node has crashed &ndash;
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/node-lifecycle.md]: femto-nit, those should be `&emdash;` :)

https://en.wikipedia.org/wiki/Dash#Em_dash

File: content/en/docs/reference/controllers/replicationcontroller.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
+---
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/replicationcontroller.md]: This document should probably be merged into the replicaset.md document and/or be deleted since this is a deprecated thing. Not in this PR, of course... just a thought for a later PR.

File: content/en/docs/reference/controllers/resource-quota.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+---
+title: Resource quota controller
+content_template: templates/concept
+---
+
+{{% capture overview %}}
+
+This controller limits the quantity of objects that can be created in a namespace
+by object type, as well as the total amount of compute resources that may be
+consumed by resources in that project.
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/resource-quota.md]: "that project" is used, but there is no mention of "project" earlier in the paragraph. This is likely due to legacy Google-specific terminology? Probably best to rephrase it here.

File: content/en/docs/reference/controllers/statefulset.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+creates these Pods *sequentially*.
+
+When the StatefulSet controller creates a Pod, it adds a label, `statefulset.kubernetes.io/pod-name`,
+that is set to the name of the Pod. This label gives each Pod a durable, stable
+network identity, that allows you to attach a Service to a specific Pod in
+the StatefulSet.
+The controller constructs the pod name label valuebased on a pattern:
+`$(statefulset name)-$(ordinal)`.
+
+### Storage
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/statefulset.md]: Similar to Deployment and Job, I think much of the remainder of this page should be in a concepts/workloads/statefulset.md document that is written from an end-user or object reference perspective. And leave this doc for implementation specifics about the controller itself.

File: content/en/docs/reference/controllers/ttl-after-finished.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+You can make changes to the TTL, i.e. to `.spec.ttlSecondsAfterFinished` field of Jobs,
+even after the resource is created or has finished.
+
+Once the TTL has expired and the Job has become eligible to delete it, the
+TTL-after-finished controller might remove that job **at any time**. This cleanup
+of expired Jobs can still happen, even if you requested a longer TTL
+and you saw a successful API response to your update request.
+
+## Time skew
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/ttl-after-finished.md]: this probably belongs in this document since it's implementation details about the controller itself.

File: content/en/docs/reference/controllers/ttl-after-finished.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+`.spec.ttlSecondsAfterFinished` field of a Job.
+See [Job](/docs/concepts/workloads/controllers/job-run-to-completion/#clean-up-finished-jobs-automatically) for more details.
+
+The TTL-after-finished controller will assume that a resource is eligible to be cleaned up
+from TTL seconds after the resource has finished, in other words, when the TTL has expired. When the
+TTL controller cleans up a resource, it will delete it cascadingly, i.e. delete
+its dependent objects together with it. Note that when the resource is deleted,
+its lifecycle guarantees, such as finalizers, will be honored.
+
+The TTL seconds can be set at any time. Here are some examples for setting the
```
Comment (jaypipes) [inline content/en/docs/reference/controllers/ttl-after-finished.md]: From here down in this section kind of feels like it belongs in a concepts/workloads/job.md document that is more oriented to the end user than the person interested in how the controller works under the hood.

File: content/en/docs/reference/controllers/attach-detach.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
+---
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/reference/controllers/attach-detach.md]: In that case, I should probably rename all the controllers based on the subdirectory they're in in the [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes/) source code.

Which I will do.

File: content/en/docs/reference/controllers/replicationcontroller.md (hunk 5 above)
Comment (sftim) [reply content/en/docs/reference/controllers/replicationcontroller.md]: There is a replication controller, and it's not the same bit of code as the ReplicaSet controller.

I would hyperlink the two but not merge them.

File: content/en/docs/reference/controllers/clusterrole-aggregation.md [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+This controller implements the `aggregationRule` property for [ClusterRoles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole), which are used in connection with
+{{< glossary_tooltip text="Role-Based Access Control" term_id="rbac" >}} (RBAC).
+
+
+{{% /capture %}}
+
+{{% capture body %}}
+
+The ClusterRole aggregation controller is built in to kube-controller-manager.
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/reference/controllers/clusterrole-aggregation.md]: AIUI it's `((ClusterRole aggregation) controller)`  rather than `(ClusterRole (aggregation controller))`. Let me know if I have plain misunderstood.

File: content/en/docs/reference/controllers/built-in.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+### Node lifecycle controller {#controller-node-lifecycle}
+
+The [node lifecycle controller](/docs/reference/controllers/node-lifecycle)
+observes the behavior of kubelet on a node, and sets (potentially also removes)
+{{< glossary_tooltip text="taints" term_id="taint" >}} on Nodes that reflect its
+findings.
+
+### Namespace lifecycle controller {#controller-namespace}
+
+When you (or any Kubernetes API client) remove a {{< glossary_tooltip term_id="namespace" >}},
```
[the remark this answers is not in the document]
Comment (lavalamp) [reply content/en/docs/reference/controllers/built-in.md]: Namespace / object is a container / containee relationship. The ordinary GC is an owner / owned relationship, which can cross object types. The namespace lifecycle controller deletes entire collections; the regular GC deletes individual objects. The namespace lifecycle controller acts on a single level relationship (even if expanded in some distant future, it'd remain a short tree); the regular GC acts on a graph.

File: content/en/docs/reference/controllers/certificate-approver.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+{{% capture overview %}}
+
+The CertificateSigningRequest approver controller is part of a set of built-in
+controllers for certificate management.
+
+{{% /capture %}}
+
+
+{{% capture body %}}
+The certificate approver is built in to kube-controller-manager.
```
[the remark this answers is not in the document]
Comment (lavalamp) [reply content/en/docs/reference/controllers/certificate-approver.md]: +1, CSR and certificates are *very* different things

**Comment by lavalamp** [comment]
My last kubecon talk was almost exactly about this: https://youtu.be/zCXiXKMqnuE

File: content/en/docs/reference/controllers/garbage-collector.md [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+Other controllers can rely on this behavior to take care of cascading deletion
+of objects via parent-child relationships.
+
+For example: if you remove a {{< glossary_tooltip term_id="deployment" >}}
+that relies on a {{< glossary_tooltip term_id="replica-set" >}} to ensure
+the right number of {{< glossary_tooltip text="Pods" term_id="pod" >}} are
+running, removing that Deployment will schedule removal of the ReplicaSet.
+
+## Owners and dependents
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/reference/controllers/garbage-collector.md]: Noted via https://github.com/kubernetes/website/issues/4510#issuecomment-511973698

File: content/en/docs/reference/controllers/horizontal-pod-autoscaler.md [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+This controller scales the number of Pods based on observed CPU utilization (or, with
+[custom metrics](https://git.k8s.io/community/contributors/design-proposals/instrumentation/custom-metrics-api.md)
+support, on some other application-provided metrics).
+
+{{% /capture %}}
+{{% capture body %}}
+
+## Automatic scaling
+
+The controller periodically adjusts the number of replicas in Deployment (or ReplicationController),
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/reference/controllers/horizontal-pod-autoscaler.md]: This page I didn't really write; it's more that https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/ is half concept page, half controller reference&mdash;and very little actual task.

I borrowed text from that Task page in a fairly copy-and-paste way, and then moved on. Maybe it needs fixing now, or maybe that can wait to a future PR. I'm fine with either choice.

File: content/en/docs/reference/controllers/node-lifecycle.md (hunk 4 above)
Comment (sftim) [reply content/en/docs/reference/controllers/node-lifecycle.md]: Maybe not a clear cut choice. See, for example, en dashes around an appositive clause in https://en.wikipedia.org/wiki/Dash#Ranges_of_values
atto-nit: the HTML entity is `&mdash;`

Using en dashes like that is a more common convention in the UK than in the US. I'll change it to better match up with the US localization.

**Comment by sftim** [comment]
What do you reckon @jaypipes? How's this now?

File: content/en/docs/reference/controllers/endpoint.md [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+{{% /capture %}}
+
+{{% capture body %}}
+
+The endpoint controller is built in to kube-controller-manager.
+
+## Controller behaviour
+
+The controller watches for Service objects. For each Service, this
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/reference/controllers/endpoint.md]: Good point

**Comment by sftim** [comment]
> My last kubecon talk was almost exactly about this: https://youtu.be/zCXiXKMqnuE

@lavalamp a good talk, too! Please do let me know if I've got any details wrong.

File: content/en/docs/reference/controllers/clusterrole-aggregation.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaypipes) [reply content/en/docs/reference/controllers/clusterrole-aggregation.md]: Ah, I see... :) Hmm, yeah, I think maybe a quick sentence explaining what is meant by "ClusterRole aggregation" might be useful. Can always be added in a followup though, of course.

File: content/en/docs/reference/controllers/replicationcontroller.md (hunk 5 above)
Comment (jaypipes) [reply content/en/docs/reference/controllers/replicationcontroller.md]: ack, hyperlink++

File: content/en/docs/reference/controllers/clusterrole-aggregation.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/reference/controllers/clusterrole-aggregation.md]: Any suggestions? It's tricky to word well.

File: content/en/docs/reference/controllers/replicationcontroller.md (hunk 5 above)
Comment (sftim) [reply content/en/docs/reference/controllers/replicationcontroller.md]: Hyperlinked

**Comment by kbhawkey** [comment]
hello, @sftim. I am in the process of reading through the files 😄 . My initial impression is that this will serve as a nice resource. Since the list of controllers is long and you have categorized some of the controllers on the `Built-in controllers` page, I would suggest to create several groupings to list on the left nav and then put the individual controllers in a category. If there are a few that do not have a single category, then list them independently. Perhaps you do not need the `Built-in Controllers` page.
The Controllers main list item in the left nav could also be listed after the API-related folder items.
Not sure about listing vendor links in the reference, for example, third party ingress controllers.

New Controllers root page preview
https://deploy-preview-15374--kubernetes-io-master-staging.netlify.com/docs/reference/controllers

File: content/en/docs/reference/controllers/built-in.md [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+## Pod management controllers {#controllers-pod-management}
+
+* [Horizontal Pod Autoscaler](/docs/reference/controllers/horizontal-pod-autoscaler/)
+* [PodDisruptionBudget controller](/docs/reference/controllers/poddisruptionbudget/)
+* [PodPreset controller](/docs/reference/access-authn-authz/admission-controllers/#podpreset)
+
+## Resource management controllers {#controllers-resource-management}
+
+* [Resource quota controller](/reference/access-authn-authz/admission-controllers/#resourcequota)
```
Comment (DanyC97) [inline content/en/docs/reference/controllers/built-in.md]: @sftim don't you need a `/docs` too ?

File: content/en/docs/reference/controllers/certificate-approver.md [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+The CSR approver is built in to kube-controller-manager.
+
+## Controller behaviour
+
+This controller acts specifically on CertificateSigningRequests (CSR) that come from
+kubelet (or that purport to come from kubelet).
+
+When kubelet is setting up on a new node, kubelet will generate a CSR and submit it
+to the Kubernetes API server using its
+[bootstrap](https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/)
```
Comment (DanyC97) [inline content/en/docs/reference/controllers/certificate-approver.md]: @sftim pls use the relative url so it does work for localization too ;)

File: content/en/docs/reference/controllers/node.md [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+controller deletes the node from its list of nodes.
+
+The third is monitoring the nodes' health. The node controller is
+responsible for updating the NodeReady condition of NodeStatus to
+ConditionUnknown when a node becomes unreachable (i.e. the node controller stops
+receiving heartbeats for some reason, e.g. due to the node being down), and then later evicting
+all the pods from the node (using graceful termination) if the node continues
+to be unreachable. (The default timeouts are 40s to start reporting
+ConditionUnknown and 5m after that to start evicting pods.) The node controller
+checks the state of each node every `--node-monitor-period` seconds.
```
Comment (DanyC97) [inline content/en/docs/reference/controllers/node.md]: @sftim imo i'd turn into a number list the the first, second and third roles, easy to read

File: content/en/docs/reference/controllers/node.md [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+The node controller looks at the state of all nodes in the cluster when making a
+decision about pod eviction.
+
+In most cases, node controller limits the eviction rate to
+`--node-eviction-rate` (default 0.1) per second, meaning it won't evict pods
+from more than 1 node per 10 seconds.
```
Comment (DanyC97) [inline content/en/docs/reference/controllers/node.md]: @sftim imo i find that we could maybe reword it a bit like 

_won't evict pods from more than 1 node on a 10 seconds window_

File: content/en/docs/reference/controllers/node.md [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+when the pods do not tolerate
+the taints.
+
+### Node leases
+
+In versions of Kubernetes prior to 1.13, NodeStatus is the heartbeat from the
+node.
+
+Starting from Kubernetes 1.13, node lease feature is introduced as an alpha feature
+(feature gate `NodeLease`, [KEP-0009](https://github.com/kubernetes/community/blob/master/keps/sig-node/0009-node-heartbeat.md)).
```
Comment (DanyC97) [inline content/en/docs/reference/controllers/node.md]: @sftim to be consistent with the rest of the urls, should we s/github.com/git.k8s.io ?

File: content/en/docs/reference/controllers/admission-pod-preset.md [hunk 22]
```diff
@@ -0,0 +1,57 @@
+---
+title: Pod preset controller
+content_template: templates/concept
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/admission-pod-preset.md]: @sftim, general comment for all pages:
I think the title text should capitalize the first letter of each major word. Also, the titles should be consistently named, contain `Controller` in the title?, and closely match the actual controller name?

File: content/en/docs/reference/controllers/admission-pod-preset.md [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+---
+
+{{% capture overview %}}
+
+The Pod preset admission controller injects configuration data into
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/admission-pod-preset.md]: - Is the controller named PodPreset or is this just the name of the admission controller flag?
- Should Admission Controller be capitalized, throughout the pages?
- Capitalize Pod/Pods?
- Since the content, in lines 21 - EOF, is taken from `https://kubernetes.io/docs/concepts/workloads/pods/podpreset/` is your intention to remove the controller-related content from the original doc, leaving the information about the resource?
- Capitalize VolumeMount?

File: content/en/docs/reference/controllers/built-in.md [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+optimizing against discovered constraints.
+{{< glossary_tooltip term_id="kubelet" >}} will update the actual state each
+time it starts or stops a scheduled Pod.
+
+Because its work is essential to Kubernetes' operation, the scheduler
+runs separately from the kube-controller-manager. This separation helps
+with control plane performance.
+
+The controllers that run inside kube-controller-manager are:
+
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/built-in.md]: Possible buckets to categorize the controllers (left nav), in no specific order?
Workload Controllers
Pod Management Controllers
Resource Management Controllers
Certificate Controllers
Storage Controllers
Networking Controllers
Cluster Orchestration(?) Controllers
Lifecycle and Garbage Collection Controllers

File: content/en/docs/reference/controllers/certificate-approver.md [hunk 25]
```diff
@@ -0,0 +1,37 @@
+---
+title: Certificate signature approver
+content_template: templates/concept
+---
+
+{{% capture overview %}}
+
+The CertificateSigningRequest approver controller is part of a set of built-in
+controllers for certificate management.
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/certificate-approver.md]: - introduce the acronym at line 8?
- what is the actual name; do you need `approver`?

File: content/en/docs/reference/controllers/daemonset.md [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+
+## Controller behavior
+
+The DaemonSet controller ensures that all eligible nodes run a copy of a Pod,
+by creating Pods with a Node assignment (one Pod for each eligible Node).
+
+This controller creates one Pod for each Node that the DaemonSet matches.
+It creates new Pods in batches, doubling the batch size each time a request
+succeeds. This helps to avoid sending a very large number of API requests thatxi
+will ultimately all fail with the same error.
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/daemonset.md]: - s/thatxi/that
- `This helps to avoid sending a very large number of API requests thatxi
will ultimately all fail with the same error`.  This could be reworded; the text seems to imply that a large number of API requests will ultimately fail for some reason.
- Future Pod creation requests are batched after the successful creation of an initial Pod/Node?

File: content/en/docs/reference/controllers/daemonset.md [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+
+To track changes efficiently, this controller labels new Pods using a hash
+of the the DaemonSet the Pod belongs to.
+
+If it's managing a Pod on a Node that no longer matches the DaemonSet,
+or where the Node no longer exists, the controller will remove that Pod.
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/daemonset.md]: What is `it`?

File: content/en/docs/reference/controllers/volume-persistentvolumeclaim.md [hunk 28]
```diff
@@ -0,0 +1,18 @@
...
+
+{{% capture overview %}}
+
+The {{< glossary_tooltip text="PersistentVolumeClaim" term_id="persistent-volume-claim" >}}
+controller is part of a set of built-in controllers for storage management.
+
+{{% /capture %}}
+
+{{% capture body %}}
+
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/volume-persistentvolumeclaim.md]: Can you add a `## Controller behavior` section so that the content is consistent with the other pages?

File: content/en/docs/reference/controllers/ttl.md [hunk 29]
```diff
@@ -0,0 +1,27 @@
+---
+title: TTL controller
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/ttl.md]: Spell out "Time-to-Live (TTL)"?

File: content/en/docs/reference/controllers/serviceaccount-token.md [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+{{% /capture %}}
+
+{{% capture body %}}
+
+The service account token controller is built in to kube-controller-manager.
+
+## Controller behaviour
+
+This controller issues API access tokens for each service account and places them into
+an associated {{< glossary_tooltip term_id="secret" >}}. Pods that can access that Secret
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/serviceaccount-token.md]: Capitalize `Service Account`?

File: content/en/docs/reference/controllers/serviceaccount-token.md [hunk 31]
```diff
@@ -0,0 +1,26 @@
+---
+title: ServiceAccount token controller
+content_template: templates/concept
+---
+
+{{% capture overview %}}
+
+One of a pair of controllers for the {{< glossary_tooltip text="ServiceAccount" term_id="service-account" >}}
+resource.
```
Comment (kbhawkey) [inline content/en/docs/reference/controllers/serviceaccount-token.md]: remove `resource`?
A ServiceAccount Token Controller generates and assigns API access tokens for a ServiceAccount.

File: content/en/docs/reference/controllers/admission-pod-preset.md (hunk 23 above)
Comment (sftim) [reply content/en/docs/reference/controllers/admission-pod-preset.md]: Thanks for the feedback.

I'm not sure that controllers have official names. I suppose there's the name of the Go package (eg `expand` or the struct (eg `expandController`).
Is it useful to mention that in the docs?

**Comment by sftim** [comment]
This should help a little with (closed) issue #10160, about documenting NodeLease.

File: content/en/docs/reference/controllers/daemonset.md (hunk 26 above)
Comment (sftim) [reply content/en/docs/reference/controllers/daemonset.md]: > a large number of API requests will ultimately fail for some reason.

File: content/en/docs/reference/controllers/admission-pod-preset.md (hunk 22 above)
Comment (sftim) [reply content/en/docs/reference/controllers/admission-pod-preset.md]: I looked at https://kubernetes.io/docs/contribute/style/ and didn't see a clear steer on whether or not to capitalize titles.

File: content/en/docs/reference/controllers/admission-pod-preset.md (hunk 23 above)
Comment (kbhawkey) [reply content/en/docs/reference/controllers/admission-pod-preset.md]: Hi @sftim. I believe there are established names for some of the controllers. It would be great to confirm the accepted names for each of the controllers. Checking the comments section for each controller _might_ help confirm the names and provide content, such as:  

`// ReplicaSetController is responsible for synchronizing ReplicaSet objects stored
// in the system with actual running pods.
type ReplicaSetController struct {`

`// TokenCleaner is a controller that deletes expired tokens
type TokenCleaner struct {`

File: content/en/docs/reference/controllers/built-in.md [hunk 32]
```diff
@@ -0,0 +1,155 @@
+---
+title: Built-in controllers
+content_template: templates/concept
+weight: 10
+---
+
+{{% capture overview %}}
+
+This page lists the {{< glossary_tooltip text="controllers" term_id="controller" >}}
```
Comment (bradtopol) [inline content/en/docs/reference/controllers/built-in.md]: When I click on the controller tooltip it takes me to the kube-controller-manager page.   But this page really doesn't have a nice description of what controllers are and there basic function.   Where will you add content that provides an introductory overview of what controllers are and what their purpose is?

File: content/en/docs/reference/controllers/built-in.md [hunk 33]
```diff
@@ -0,0 +1,155 @@
...
+desired set of running Pods and match that against the available Nodes,
+optimizing against discovered constraints.
+{{< glossary_tooltip term_id="kubelet" >}} will update the actual state each
+time it starts or stops a scheduled Pod.
+
+Because its work is essential to Kubernetes' operation, the scheduler
+runs separately from the kube-controller-manager. This separation helps
+with control plane performance.
+
+The controllers that run inside kube-controller-manager are:
```
Comment (bradtopol) [inline content/en/docs/reference/controllers/built-in.md]: The transition on the page from "The controllers that run inside kube-controller-manager are:" to then going to a bold heading of "Controllers for running workloads on Kubernetes" looks really weird.  Check out https://kubernetes.io/docs/setup/ which shows a nice transition from a list of topics (Learning Environment, Production environment) to their sections below the introductory paragraph

File: content/en/docs/reference/controllers/built-in.md (hunk 32 above)
Comment (sftim) [reply content/en/docs/reference/controllers/built-in.md]: I have a draft for an initial [overview page about Controllers](https://deploy-preview-15373--kubernetes-io-master-staging.netlify.com/docs/concepts/architecture/controller/) as part of PR #15373.

and I plan to propose that in a separate pull request. I'd also change the _controller_ glossary entry to point to that new page.

File: content/en/docs/reference/controllers/built-in.md (hunk 33 above)
Comment (sftim) [reply content/en/docs/reference/controllers/built-in.md]: Thanks for your feedback on this.

File: content/en/docs/reference/controllers/attach-detach.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/reference/controllers/attach-detach.md]: I have a better idea. I will rename them based on the names visible to cluster operators, as per `--controllers` in https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/

(these must also be in the source code, so I'll see if I can spot how that's done)

File: content/en/docs/reference/controllers/admission-pod-preset.md (hunk 23 above)
Comment (sftim) [reply content/en/docs/reference/controllers/admission-pod-preset.md]: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/#options lists controller names (in the `--controllers` parameter) that are visible to cluster operators.

Now I've seen that, I'm minded to put those lower-case names somewhere in the page for each controller.

**Comment by sftim** [comment]
Something that occurred to me a little while ago: possibly, this documentation could live as reference documentation in [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes)

I'll continue with this approach for now, if people reading this like the above idea though I can look into taking a different tack.

**Comment by sftim** [comment]
Didn't get round to this yet. I still plan to!

File: content/en/docs/reference/controllers/built-in.md (hunk 32 above)
Comment (sftim) [reply content/en/docs/reference/controllers/built-in.md]: The separate pull request about a controller concept is #15733

**Comment by zacharysarah** [comment]
@sftim Please feel free to reopen this when you're ready!

/close

**Comment by sftim** [comment]
/retitle Add reference documentation for built-in controllers

File: content/en/docs/reference/controllers/built-in.md (hunk 32 above)
Comment (sftim) [reply content/en/docs/reference/controllers/built-in.md]: Merged #15733

File: content/en/docs/reference/controllers/serviceaccount-token.md (hunk 30 above)
Comment (sftim) [reply content/en/docs/reference/controllers/serviceaccount-token.md]: Done!

**Comment by sftim** [comment]
Rebased & tweaked
/reopen
